pkg/activator/throttler.go

```diff
@@ -0,0 +1,169 @@
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"sync"
+)
+
+const capacityUpdateFailure = "updating capacity failed"
+
+// NewThrottler creates a new Throttler
+func NewThrottler(breakerParams queue.BreakerParams, revisionInformer v1alpha1.RevisionInformer, logger *zap.SugaredLogger) *Throttler {
+	breakers := make(map[RevisionID]*queue.Breaker)
+	return &Throttler{breakers: breakers, breakerParams: breakerParams, revisionInformer: revisionInformer, logger: logger}
+}
+
+// Throttler keeps the mapping of Revisions to Breakers
+// and allows updating max concurrency dynamically of respective Breakers.
+// Max concurrency is essentially the number of semaphore tokens the Breaker has in rotation.
+// The manipulation of the parameter is done via `UpdateCapacity()` method.
+// It enables the use case to start with max concurrency set to 0 (no requests are sent because no endpoints are available)
+// and gradually increase its value depending on the external condition (e.g. new endpoints become available)
+type Throttler struct {
+	breakers         map[RevisionID]*queue.Breaker
+	revisionInformer v1alpha1.RevisionInformer
+	breakerParams    queue.BreakerParams
+	logger           *zap.SugaredLogger
+	mux              sync.Mutex
+}
+
+// Remove deletes the breaker from the bookkeeping
+func (t *Throttler) Remove(rev RevisionID) {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	delete(t.breakers, rev)
+}
+
+// UpdateCapacity updates the max concurrency of the Breaker corresponding to a revision.
+// It create a new breaker and saves into our bookkeeping if doesn't exist.
+// This is important for not loosing the update signals
+// for requests that come before the requests reached the Activator's Handler.
+func (t *Throttler) UpdateCapacity(rev RevisionID, size int32) error {
+	t.mux.Lock()
+	defer t.mux.Unlock()
+	revision, exists, err := t.getRevision(rev.Namespace, rev.Name, t.revisionInformer)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return fmt.Errorf("no revision was found for: %s", rev)
+	}
+	cc := int32(revision.Spec.ContainerConcurrency)
+
+	if cc == 0 {
+		cc = t.breakerParams.MaxConcurrency
+	}
+	breaker, ok := t.breakers[rev]
+	if !ok {
+		breaker = queue.NewBreaker(t.breakerParams)
+		t.breakers[rev] = breaker
+	}
+	delta := size*cc - breaker.Capacity()
+	// Either the same number of endpoints received or
+	// some endpoints that were not registered at this Activator are torn down.
+	// In both cases do not perform an update.
+	if delta == 0 || (delta < 0 && breaker.Capacity() == 0) {
+		return nil
+	}
+	if err := breaker.UpdateConcurrency(delta); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Try executes the `function` on the Breaker if it has enough capacity.
+// The function returns an error otherwise.
+func (t *Throttler) Try(rev RevisionID, function func()) error {
+	t.mux.Lock()
+	breaker, ok := t.breakers[rev]
+	if !ok {
+		breaker = queue.NewBreaker(t.breakerParams)
+		t.breakers[rev] = breaker
+	}
```

Comment:
Yes, we need to create the Breaker here - in case if the request came in before the Endpoints got updated. So we have to leave with it for now I think, on the a side note - the contention will only take place for the first 7-8 seconds before the first node showed up.

Reply:
Right, I guess the contention is not ultra critical because latency is delayed anyway. Things like this can be solved after the fact once they prove to be an issue.